src/rulesets/LinearAlgebra/dense.jl

```diff
@@ -118,6 +118,113 @@ function rrule(::typeof(*), A::AbstractMatrix{<:Real}, B::AbstractMatrix{<:Real}
     return A * B, times_pullback
 end
 
+#####
+##### `pinv`
+#####
+
+@scalar_rule pinv(x) -(Ω ^ 2)
+
+function frule(
+    (_, Δx),
+    ::typeof(pinv),
+    x::AbstractVector{T},
+    tol::Real = 0,
+) where {T<:Union{Real,Complex}}
+    y = pinv(x, tol)
+    ∂y′ = sum(abs2, parent(y)) .* Δx .- 2real(y * Δx) .* parent(y)
+    ∂y = y isa Transpose ? transpose(∂y′) : adjoint(∂y′)
+    return y, ∂y
+end
+
+function frule(
+    (Δself, Δx),
+    ::typeof(pinv),
+    x::LinearAlgebra.AdjOrTransAbsVec{T},
+    tol::Real = 0,
+) where {T<:Union{Real,Complex}}
+    x′ = parent(x)
+    if x isa Transpose
+        x′, Δx′ = conj(x′), conj(transpose(Δx))
+    else
+        Δx′ = adjoint(Δx)
+    end
+    y′, ∂y′ = frule((Δself, Δx′, Zero()), pinv, x′, tol)
+    y, ∂y = parent(y′), parent(∂y′)
+    return y, ∂y
+end
+
+# Formula for derivative adapted from Eq 4.12 of
+# Golub, Gene H., and Victor Pereyra. "The Differentiation of Pseudo-Inverses and Nonlinear
+# Least Squares Problems Whose Variables Separate."
+# SIAM Journal on Numerical Analysis 10(2). (1973). 413-432. doi: 10.1137/0710036
+function frule((_, ΔA), ::typeof(pinv), A::AbstractMatrix{T}; kwargs...) where {T}
+    Y = pinv(A; kwargs...)
+    m, n = size(A)
+    # contract over the largest dimension
+    if m ≤ n
+        ∂Y = -Y * (ΔA * Y)
+        _add!(∂Y, (ΔA' - Y * (A * ΔA')) * (Y' * Y)) # (I - Y A) ΔA' Y' Y
+        _add!(∂Y, Y * (Y' * ΔA') * (I - A * Y)) # Y Y' ΔA' (I - A Y)
+    else
+        ∂Y = -(Y * ΔA) * Y
+        _add!(∂Y, (I - Y * A) * (ΔA' * Y') * Y) # (I - Y A) ΔA' Y' Y
+        _add!(∂Y, (Y * Y') * (ΔA' - (ΔA' * A) * Y)) # Y Y' ΔA' (I - A Y)
+    end
+    return Y, ∂Y
+end
+
+function rrule(
+    ::typeof(pinv),
+    x::AbstractVector{T},
+    tol::Real = 0,
+) where {T<:Union{Real,Complex}}
+    y = pinv(x, tol)
+    function pinv_pullback(Δy)
+        ∂x = sum(abs2, parent(y)) .* vec(Δy') .- 2real(y * Δy') .* parent(y)
+        return (NO_FIELDS, ∂x, Zero())
+    end
+    return y, pinv_pullback
```

Comment:
```suggestion
    function _pinv_pullback(Δy)
        ∂x = sum(abs2, parent(y)) .* vec(Δy') .- 2real(y * Δy') .* parent(y)
        return (NO_FIELDS, ∂x, Zero())
    end
    
    y = pinv(x, tol)
    
    return y, _pinv_pullback
```

Reply:
The two changes I see here are declaring `y` after the pullback and changing the pullback name. The first change seems less readable, as one needs to scroll down to find `y`, and most of the existing rules in ChainRules declare `y` first when it is used. The second change does not conform with the ChainRules convention (namely, preferring to name pullbacks `XXX_pullback`, where `XXX` is the name of the primal function.